mesonbuild/wrap/wrap.py

```diff
@@ -67,93 +67,103 @@ def open_wrapdburl(urlstring):
-                v = v.strip()
-                self.values[k] = v
+        self.basename = os.path.basename(fname)
+        try:
+            self.config = configparser.ConfigParser()
+            self.config.read(fname)
+        except:
+            raise WrapException('Failed to parse {}'.format(self.basename))
+        if len(self.config.sections()) < 1:
+            raise WrapException('Missing sections in {}'.format(self.basename))
+        self.wrap_section = self.config.sections()[0]
+        if not self.wrap_section.startswith('wrap-'):
+            m = '{!r} is not a valid first section in {}'
+            raise WrapException(m.format(self.wrap_section, self.basename))
+        self.type = self.wrap_section[5:]
+        self.values = dict(self.config[self.wrap_section])
 
     def get(self, key):
-        return self.values[key]
+        try:
+            return self.values[key]
+        except KeyError:
+            m = 'Missing key {!r} in {}'
+            raise WrapException(m.format(key, self.basename))
 
     def has_patch(self):
         return 'patch_url' in self.values
 
 class Resolver:
-    def __init__(self, subdir_root, wrap_mode=WrapMode(1)):
+    def __init__(self, subdir_root, wrap_mode=WrapMode.default):
         self.wrap_mode = wrap_mode
         self.subdir_root = subdir_root
         self.cachedir = os.path.join(self.subdir_root, 'packagecache')
 
     def resolve(self, packagename):
-        # Check if the directory is already resolved
-        dirname = Path(os.path.join(self.subdir_root, packagename))
-        subprojdir = os.path.join(*dirname.parts[-2:])
-        if dirname.is_dir():
-            if (dirname / 'meson.build').is_file():
-                # The directory is there and has meson.build? Great, use it.
-                return packagename
-            # Is the dir not empty and also not a git submodule dir that is
-            # not checkout properly? Can't do anything, exception!
-            elif next(dirname.iterdir(), None) and not (dirname / '.git').is_file():
-                m = '{!r} is not empty and has no meson.build files'
-                raise RuntimeError(m.format(subprojdir))
-        elif dirname.exists():
-            m = '{!r} already exists and is not a dir; cannot use as subproject'
-            raise RuntimeError(m.format(subprojdir))
-
-        dirname = str(dirname)
+        self.packagename = packagename
+        self.directory = packagename
+        # We always have to load the wrap file, if it exists, because it could
+        # override the default directory name.
+        p = self.load_wrap()
+        if p and 'directory' in p.values:
+            self.directory = p.get('directory')
+        dirname = os.path.join(self.subdir_root, self.directory)
+        meson_file = os.path.join(dirname, 'meson.build')
+
+        # The directory is there and has meson.build? Great, use it.
+        if os.path.exists(meson_file):
+            return self.directory
```

Comment:
`wrap-git` with `revision = head` is documented to repull on each meson run. I think vcs·es need to have code here that can do that. (svn and hg seem to have similar behavior)

(It seems in previous code that did only work when `directory` in the wrap was set to a different value than the subproject name)

Reply:
I didn't know that was documented, but that never worked. I personally think it should not do it, could we just update the doc then? Where is that written?

Reply:
http://mesonbuild.com/Wrap-dependency-system-manual.html#branching-subprojects-directly-from-git
> The first is head which will cause Meson to track the master head (doing a repull whenever the build definition is altered).

Removing those needs discussion and should be another PR ideally. Also such changes that do not fit the PR title should be called out explicitly in the PR description/discussion if they for some reason are not separable.

Reply:
Yes, I agree we should do something about that in a separate PR. Since that was already broken, I think this PR is fine, we can do a follow-up branch later. Right?

Reply:
Opened PR #4389 for that.